### PR TITLE
Keep source data into output data with the same classes as inputs

### LINKED ISSUE #190: keep respects original value more.
https://github.com/nlmixr2/nlmixr2est/issues/51

### PR DESCRIPTION
**Pull request by billdenney**
@mattfidler, hopefully this helps fix #190 at least a bit.

### CODE REVIEW

**Comment by mattfidler** [comment]
Thanks.

**Comment by billdenney** [comment]
Add `rxRowNum_` to the input data here:

https://github.com/nlmixr2/rxode2/blob/2c13d147fce9aa9e5fc31db802ce13b59f53e25d/R/rxsolve.R#L1453-L1496

Keep it in the input dataset, then merge back in the keep columns here:

https://github.com/nlmixr2/rxode2/blob/2c13d147fce9aa9e5fc31db802ce13b59f53e25d/R/rxsolve.R#L1559-L1607

File: R/rxsolve.R [hunk 1]
```diff
@@ -1617,7 +1651,17 @@ rxSolve.default <- function(object, params = NULL, events = NULL, inits = NULL,
   } else if (.ctl$returnType == 5L) {
     .ret <- tibble::as_tibble(.ret)
   }
-  return(.ret)
+  # Restore the keep columns, if applicable
+  if (length(paramsEvents$keep) > 0) {
+    # Use a bespoke merge to do a quick, simple merge
```
Comment (billdenney) [inline R/rxsolve.R:1656]: @mattfidler, here is where to make the merge keep the class.

**Comment by billdenney** [comment]
@mattfidler, this is ready for your final touches, I think.

File: R/rxsolve.R [hunk 2]
```diff
@@ -1320,15 +1404,6 @@ rxSolve.default <- function(object, params = NULL, events = NULL, inits = NULL,
 
     }
   }
-  .n1 <- setdiff(intersect(tolower(names(params)), tolower(names(.ctl$iCov))), "id")
```
Comment (mattfidler) [inline R/rxsolve.R:1323]: This is no longer included (and should be)....

Comment (mattfidler) [reply R/rxsolve.R:1323]: Nevermind I see it now.

Comment (billdenney) [reply R/rxsolve.R:1323]: Yeah, I also tried to simplify the overall dataset generation functions.

**Comment by mattfidler** [comment]
```sh
==17639== Memcheck, a memory error detector
==17639== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==17639== Using Valgrind-3.20.0 and LibVEX; rerun with -h for copyright info
==17639== Command: /usr/lib/R/bin/exec/R
==17639== 

R version 4.2.2 Patched (2022-11-10 r83330) -- "Innocent and Trusting"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> devtools::test(filter="backward")
ℹ Testing rxode2
==17639== Source and destination overlap in memcpy_chk(0x1ffefb8044, 0x1ffefb80a8, 897)
==17639==    at 0x4E10570: __memcpy_chk (vg_replace_strmem.c:1741)
==17639==    by 0x5176F8C: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517C4A9: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517D1E4: Rf_yyparse (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517F9B5: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51802ED: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5180722: R_ParseVector (in /usr/lib/R/lib/libR.so)
==17639==    by 0x521A19E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Source and destination overlap in memcpy_chk(0x1ffefb8044, 0x1ffefb80a8, 897)
==17639==    at 0x4E10570: __memcpy_chk (vg_replace_strmem.c:1741)
==17639==    by 0x5177BBA: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517C4A9: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517D1E4: Rf_yyparse (in /usr/lib/R/lib/libR.so)
==17639==    by 0x517F9B5: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51802ED: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5180722: R_ParseVector (in /usr/lib/R/lib/libR.so)
==17639==    by 0x521A19E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639== 
rxode2 2.0.11.9000 using 4 threads (see ?getRxThreads)
✔ | F W S  OK | Context
⠸ | 2       2 | backward                                                                                                               ==17639== Invalid read of size 4
==17639==    at 0x5166742: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x1913F056: Rcpp_eval_impl (Rcpp_eval.h:49)
==17639==    by 0x1913F056: Rcpp::Rcpp_eval(SEXPREC*, SEXPREC*) (Rcpp_eval.h:92)
==17639==    by 0x1913FDF4: Rcpp_fast_eval (Rcpp_eval.h:69)
==17639==    by 0x1913FDF4: get_last_call (exceptions.h:272)
==17639==    by 0x1913FDF4: SEXPREC* exception_to_condition_template<std::exception>(std::exception const&, bool) (exceptions.h:335)
==17639==    by 0x19127635: exception_to_r_condition (exceptions.h:353)
==17639==    by 0x19127635: _rxode2et_et_.cold (RcppExports.cpp:48)
==17639==    by 0x510E8D9: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x515077D: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x126A804C: magrittr_pipe (in /usr/lib/R/site-library/magrittr/libs/magrittr.so)
==17639==    by 0x510CDE7: ??? (in /usr/lib/R/lib/libR.so)
==17639==  Address 0x1ffef40508 is on thread 1's stack
==17639==  152952 bytes below stack pointer
==17639== 
==17639== Invalid read of size 4
==17639==    at 0x51690A0: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x1913F056: Rcpp_eval_impl (Rcpp_eval.h:49)
==17639==    by 0x1913F056: Rcpp::Rcpp_eval(SEXPREC*, SEXPREC*) (Rcpp_eval.h:92)
==17639==    by 0x1913FDF4: Rcpp_fast_eval (Rcpp_eval.h:69)
==17639==    by 0x1913FDF4: get_last_call (exceptions.h:272)
==17639==    by 0x1913FDF4: SEXPREC* exception_to_condition_template<std::exception>(std::exception const&, bool) (exceptions.h:335)
==17639==    by 0x19127635: exception_to_r_condition (exceptions.h:353)
==17639==    by 0x19127635: _rxode2et_et_.cold (RcppExports.cpp:48)
==17639==    by 0x510E8D9: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x515077D: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x126A804C: magrittr_pipe (in /usr/lib/R/site-library/magrittr/libs/magrittr.so)
==17639==  Address 0x1ffef40508 is on thread 1's stack
==17639==  152856 bytes below stack pointer
==17639== 
==17639== Invalid read of size 4
==17639==    at 0x50F5F88: Rf_begincontext (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5168034: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x1913F056: Rcpp_eval_impl (Rcpp_eval.h:49)
==17639==    by 0x1913F056: Rcpp::Rcpp_eval(SEXPREC*, SEXPREC*) (Rcpp_eval.h:92)
==17639==    by 0x1913FDF4: Rcpp_fast_eval (Rcpp_eval.h:69)
==17639==    by 0x1913FDF4: get_last_call (exceptions.h:272)
==17639==    by 0x1913FDF4: SEXPREC* exception_to_condition_template<std::exception>(std::exception const&, bool) (exceptions.h:335)
==17639==    by 0x19127635: exception_to_r_condition (exceptions.h:353)
==17639==    by 0x19127635: _rxode2et_et_.cold (RcppExports.cpp:48)
==17639==    by 0x510E8D9: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x515077D: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==  Address 0x1ffef40678 is on thread 1's stack
==17639==  151848 bytes below stack pointer
==17639== 
==17639== Conditional jump or move depends on uninitialised value(s)
==17639==    at 0x50F6B31: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516BD96: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Use of uninitialised value of size 8
==17639==    at 0x50F6B27: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516BD96: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Use of uninitialised value of size 8
==17639==    at 0x50F6B1B: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516BD96: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Invalid read of size 8
==17639==    at 0x50F6B27: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516BD96: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5149664: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51666EF: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516827E: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51690C4: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==17639==    by 0x516681B: Rf_eval (in /usr/lib/R/lib/libR.so)
==17639==  Address 0x7c8d9e00000 is not stack'd, malloc'd or (recently) free'd
==17639== 

 *** caught segfault ***
address 0x7c8d9e00000, cause 'memory not mapped'
==17639== Conditional jump or move depends on uninitialised value(s)
==17639==    at 0x5137378: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Conditional jump or move depends on uninitialised value(s)
==17639==    at 0x513737D: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Conditional jump or move depends on uninitialised value(s)
==17639==    at 0x513738C: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Use of uninitialised value of size 8
==17639==    at 0x5137373: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Use of uninitialised value of size 8
==17639==    at 0x5137383: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639== 
==17639== Invalid read of size 4
==17639==    at 0x5137373: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639==  Address 0x7c8d9e00008 is not stack'd, malloc'd or (recently) free'd
==17639== 
==17639== 
==17639== Process terminating with default action of signal 11 (SIGSEGV)
==17639==  Access not within mapped region at address 0x7C8D9E00008
==17639==    at 0x5137373: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x51374B6: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x5199618: ??? (in /usr/lib/R/lib/libR.so)
==17639==    by 0x551B51F: ??? (in /usr/lib/x86_64-linux-gnu/libc.so.6)
==17639==    by 0x50F6B26: ??? (in /usr/lib/R/lib/libR.so)
==17639==  If you believe this happened as a result of a stack
==17639==  overflow in your program's main thread (unlikely but
==17639==  possible), you can try to increase the size of the
==17639==  main thread stack using the --main-stacksize= flag.
==17639==  The main thread stack size used in this run was 8388608.
==17639== 
==17639== HEAP SUMMARY:
==17639==     in use at exit: 224,481,495 bytes in 61,414 blocks
==17639==   total heap usage: 734,571 allocs, 673,157 frees, 1,908,628,738 bytes allocated
==17639== 
==17639== LEAK SUMMARY:
==17639==    definitely lost: 0 bytes in 0 blocks
==17639==    indirectly lost: 0 bytes in 0 blocks
==17639==      possibly lost: 212,586 bytes in 6,456 blocks
==17639==    still reachable: 224,268,909 bytes in 54,958 blocks
==17639==                       of which reachable via heuristic:
==17639==                         newarray           : 4,264 bytes in 1 blocks
==17639==         suppressed: 0 bytes in 0 blocks
==17639== Rerun with --leak-check=full to see details of leaked memory
==17639== 
==17639== Use --track-origins=yes to see where uninitialised values come from
==17639== For lists of detected and suppressed errors, rerun with: -s
==17639== ERROR SUMMARY: 47 errors from 15 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)
```

**Comment by mattfidler** [comment]
This still breaks something... I'm a bit stumped.  Now that I am trying this, I recall trying this approach before (perhaps I am imagining it).

**Comment by mattfidler** [comment]
Thanks for your help.  Im unsure why this is breaking but the c approach works fine.

**Pull request closed without merging**